### PR TITLE
[Clang] Allow all address spaces to be converted to the default

### LINKED ISSUE #112233: [Clang] Using address spaces in C/C++ assumes unspecified addresses are private instead of generic
The OpenCL 2.0 rules allow us to implicitly convert specific address spaces into generic address spaces. This is also how languages like CUDA or OpenMP work. Using C/C++ for targeting the GPU should also allow this, but it is currently unnecessarily restrictive. See the example in https://godbolt.org/z/oa5Tob8v6.

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
Currently, we want to use the OpenCL attributes to indicate the address
space. Languages like SYCL, OpenMP, HIP, and CUDA allow their address
space qualifiers to be implicitly converted to generic, as does CL2.0
(except for __constant). We want this behavior when targeting C/C++
directly with the OpenCL attributes or when using CUDA/OpenMP and want to
qualify pointers with the types. The current CL1.0 rules are
unnecessarily strict when the GPU targets are expected to handle flat
pointers.

This patch changes the logic to allow any cast if the target is Generic.
For OpenCL every global will have `opencl_generic` or `opencl_private`
attributes unless it's some kind of function object.

I'm not sure if this is the best and most correct solution. If we want
to leave OpenCL untouched we could just check the language before
checking the rules instead of just the address spaces. Alternatively, we
could make an entirely new set of address space attributes that drops
the `opencl` name and use those (but then we'd need to duplicate the
same sema checking everywhere).

Fixes: https://github.com/llvm/llvm-project/issues/112233

### CODE REVIEW

File: clang/include/clang/AST/Type.h [hunk 1]
```diff
@@ -721,7 +721,9 @@ class Qualifiers {
            // to implicitly cast into the default address space.
            (A == LangAS::Default &&
             (B == LangAS::cuda_constant || B == LangAS::cuda_device ||
-             B == LangAS::cuda_shared));
+             B == LangAS::cuda_shared)) ||
+           // Otherwise, assume the default address space is compatible.
+           (A == LangAS::Default);
```
Comment (arsenm) [inline clang/include/clang/AST/Type.h:726]: Default is kind of broken, at least for OpenCL. Should avoid attaching any behavior to it

Comment (jhuber6) [reply clang/include/clang/AST/Type.h:726]: Yeah I noted  there's actually a case where OpenCL touches this for function pointers. I might just have the OpenCL language call a different version of this. Overall the other targets just allow any AS to decay to default and that's likely the behavior we want when not bound by OpenCL semantics.

**Comment by shiltian** [comment]
I think the default AS is expected to be losslessly converted to other AS but not the other way around, though I understand it is not clearly stated in LangRef or other places and a lot of code just assumes it can.

**Comment by jhuber6** [comment]
> I think the default AS is expected to be losslessly converted to other AS but not the other way around, though I understand it is not clearly stated in LangRef or other places and a lot of code just assumes it can.

All the other language address spaces (cuda, OpenMP, sycl) support converting to the default address space. OpenCL supports it except for constant.

**Comment by AlexVlx** [comment]
> I think the default AS is expected to be losslessly converted to other AS but not the other way around, though I understand it is not clearly stated in LangRef or other places and a lot of code just assumes it can.

I don't think this expectation is thoroughly encoded anywhere, and it's definitely not binding today (there are targets who have a mapping of default that breaks this assumption). Whilst the premise is interesting I'm not sure this patch is doing the right thing. If you are in vanilla C/C++, you wouldn't have anything but default, and hopefully the default is mapped to something sane on the target. 

Running into an observable situation where this is a concern means that either you've messed around with (non C/C++) attributes, or are linking in something exciting. Neither of which constitutes valid C/C++, and I'm leaning towards saying shouldn't be allowed to silently work - we probably should refrain from infecting C/C++ with explicit address spaces a la OpenCL. What are you actually running into @jhuber6 where this is a concern?

**Comment by jhuber6** [comment]
> Running into an observable situation where this is a concern means that either you've messed around with (non C/C++) attributes, or are linking in something exciting. Neither of which constitutes valid C/C++, and I'm leaning towards saying shouldn't be allowed to silently work - we probably should refrain from infecting C/C++ with explicit address spaces a la OpenCL. What are you actually running into @jhuber6 where this is a concern? More specifically, you are saying "we want to do X, in C++" -> why?

I'm trying to port the OpenMP device runtime to just use C/C++, we use a lot of `__local` variables there which this turns into a mess of C-style casts. It also interacts badly with the `this` pointer if you put a class inside of anything with an address space. C/C++ says nothing about address spaces, but these basically act as target specific attributes on certain types.

This is a case where it's actually harder to do something in C/C++ than OpenCL and I don't want it to be that way. See https://godbolt.org/z/1Gn71qqPT for an example of what I'm talking about.

TL;DR, every other language allows this and I want to be able to use classes or LDS variables without spamming `(int &)` everywhere.

**Comment by AlexVlx** [comment]
> > Running into an observable situation where this is a concern means that either you've messed around with (non C/C++) attributes, or are linking in something exciting. Neither of which constitutes valid C/C++, and I'm leaning towards saying shouldn't be allowed to silently work - we probably should refrain from infecting C/C++ with explicit address spaces a la OpenCL. What are you actually running into @jhuber6 where this is a concern? More specifically, you are saying "we want to do X, in C++" -> why?
> 
> I'm trying to port the OpenMP device runtime to just use C/C++, we use a lot of `__local` variables there which this turns into a mess of C-style casts. It also interacts badly with the `this` pointer if you put a class inside of anything with an address space. C/C++ says nothing about address spaces, but these basically act as target specific attributes on certain types.
> 
> This is a case where it's actually harder to do something in C/C++ than OpenCL and I don't want it to be that way. See https://godbolt.org/z/1Gn71qqPT for an example of what I'm talking about.
> 
> TL;DR, every other language allows this and I want to be able to use classes or LDS variables without spamming `(int &)` everywhere.

Well it's harder to do OpenCL in C++ than doing OpenCL in OpenCL is not entirely surprising, is it? I understand you do not want to do it that way, however this allows for what is essentially non-standard C++ to sneakily start "working" as if it were standard C++. Every other language in this case == every other GPU offload language modelled after CUDA/OpenCL, but I don't know if that should mean we can simply drop something in an otherwise standard C++ compilation. This is important because these are typefully significant target specific attributes, with consequential semantics in the AST (for example they contribute to overload resolution). It's also plausibly important for quirky cases such as bit-patterns `nullptr` / `NULL` for different address spaces - as I said, not all AS-casts to generic have to be NOPs. So overall I'm leaning against doing it as is; guarding it under some flavour of `allow-relaxed-address-space-handling` or something that sounds better but is still noisy enough to make it clear there's danger seems preferable, at least to me.

**Comment by jhuber6** [comment]
> So overall I'm leaning against doing it as is; guarding it under some flavour of `allow-relaxed-address-space-handling` or something that sounds better but is still noisy enough to make it clear there's danger seems preferable, at least to me.

A language option is reasonable if we don't want default behavior, though it does add a bit of magic to what is otherwise straightfoward on all the other standard targets.

File: clang/include/clang/AST/Type.h (hunk 1 above)
Comment (bader) [reply clang/include/clang/AST/Type.h:726]: @jhuber6, I agree with you, but you might want to check with @AnastasiaStulova.

We had a couple of lengthy discussions about re-using OpenCL attributes in SYCL mode ([here](https://reviews.llvm.org/D80932) and [here](https://github.com/intel/llvm/pull/1039/files)), but if I recall it correctly the conclusion was that OpenCL attributes inherit OpenCL semantics in non-OpenCL modes as well. The solution for SYCL mode was adding new attributes ([review](https://reviews.llvm.org/D89909)).

@Naghasan, FYI.

**Comment by arsenm** [comment]
> I'm thinking we could have a language option like `-fimplicit-default-addrspace-cast` that allows all conversions to default. Then if we want users to be aware that this even exists we could suggest it in the error message if it would fix it.
> 
> I do understand that there's some weird semantics here, and it does change some surprisingly unrelated tests so I don't think the patch as-is will be sound. Hiding it behind an option makes sense so I can do that.

I think we should avoid modality / language dependence. I thought the problem you were dealing with was the cuda_ attributes not behaving consistently in C++ mode?

**Comment by arsenm** [comment]
> They should not "just work" if you're compiling for vanilla C/C++ (so not CUDA C++, or SYCL, or OpenCL C/C++ etc.) because neither of those defines such a notion.

a6a237f2046ad8993db30481c8b61aeb2f73a5ad added an explicit addrspace_cast operator. It would be better to use something like this rather than just freely enabling any implicit cast

**Comment by jhuber6** [comment]
> > They should not "just work" if you're compiling for vanilla C/C++ (so not CUDA C++, or SYCL, or OpenCL C/C++ etc.) because neither of those defines such a notion.
> 
> [a6a237f](https://github.com/llvm/llvm-project/commit/a6a237f2046ad8993db30481c8b61aeb2f73a5ad) added an explicit addrspace_cast operator. It would be better to use something like this rather than just freely enabling any implicit cast

That's only enabled in OpenCLC++ I think? We could probably make a patch to enable it in C++. (Guessing we'd need to call it `__addrspace_cast` or `__builtin_addrspace_cast`. However it's still a huge pain to work with.

I tried adding an option to relax the rules but it's much more difficult than I anticipated because it requires forwarding  the language options to every 100 or so places that Sema wants to check if the types are legal. OpenCL3.0 lets you enable it for convenience so I'd like that kind of behavior (We probably also do want addrspace cast in general). But it seems like it might be really, really annoying to enable that without becoming beholden to OpenCL rules. The only way I could thing to change these rules optionally would be to introduce a ton of new addrspace rules which are basically just OpenCL but not.

File: clang/include/clang/AST/Type.h (hunk 1 above)
Comment (Naghasan) [reply clang/include/clang/AST/Type.h:726]: > that's likely the behavior we want when not bound by OpenCL semantics.

I kind of agree, but that's up to the target to tell if this makes sense IMO. My mental model of what `Default` means in the case of CUDA/HIP/SYCL/OpenMP is "the address of a variable maps to the flat address space of the target".

My understanding here is you want a bypass for target address spaces to decay into default. While I think this is desirable in general, this disregards potential target limitations. The one I have in mind is for SPIR/SPIR-V, the target address space for constant shouldn't be allowed to decay into `Default`. There is some related discussions [here](https://discourse.llvm.org/t/rfc-improved-address-space-conversion-semantics-for-targets-and-language-dialects/51365) as well.
